crates/ink/macro/src/event/mod.rs

```diff
@@ -0,0 +1,207 @@
+    let mut topic_err: Option<syn::Error> = None;
+    s.variants_mut()[0].filter(|bi| {
+        match has_ink_topic_attribute(&bi.ast().attrs) {
+            Ok(has_attr) => has_attr,
+            Err(err) => {
+                match topic_err {
+                    Some(ref mut topic_err) => topic_err.combine(err),
+                    None => topic_err = Some(err),
+                }
+                false
+            }
+        }
+    });
+    if let Some(err) = topic_err {
+        return Err(err)
+    }
+
+    let variant = &s.variants()[0];
+
+    // Anonymous events require 1 fewer topics since they do not include their signature.
+    let anonymous_topics_offset = usize::from(!anonymous);
+    let len_topics = variant.bindings().len() + anonymous_topics_offset;
+
+    let remaining_topics_ty = match len_topics {
+        0 => quote_spanned!(span=> ::ink::env::event::state::NoRemainingTopics),
+        _ => {
+            quote_spanned!(span=> [::ink::env::event::state::HasRemainingTopics; #len_topics])
+        }
+    };
+
+    let event_ident = variant.ast().ident;
+    let signature_topic = if !anonymous {
+        let topic_str = signature_topic(variant.ast().fields, event_ident);
+        quote_spanned!(span=> ::core::option::Option::Some(::ink::blake2x256!(#topic_str)))
+    } else {
+        quote_spanned!(span=> ::core::option::Option::None)
+    };
+    let event_signature_topic = if anonymous {
+        None
+    } else {
+        Some(quote_spanned!(span=>
+            .push_topic(Self::SIGNATURE_TOPIC.as_ref())
+        ))
+    };
+
+    let topics = variant.bindings().iter().fold(quote!(), |acc, field| {
+        let field_ty = &field.ast().ty;
+        let field_span = field_ty.span();
+        quote_spanned!(field_span=>
+            #acc
+            .push_topic(::ink::as_option!(#field))
+        )
+    });
+    let pat = variant.pat();
+    let topics_builder = quote!(
+        #pat => {
+            builder
+                .build::<Self>()
+                #event_signature_topic
+                #topics
+                .finish()
+        }
+    );
+
+    Ok(s.bound_impl(quote!(::ink::env::Event), quote! {
+        type RemainingTopics = #remaining_topics_ty;
+
+        const SIGNATURE_TOPIC: ::core::option::Option<[::core::primitive::u8; 32]> = #signature_topic;
+
+        fn topics<E, B>(
+            &self,
+            builder: ::ink::env::event::TopicsBuilder<::ink::env::event::state::Uninit, E, B>,
+        ) -> <B as ::ink::env::event::TopicsBuilderBackend<E>>::Output
+        where
+            E: ::ink::env::Environment,
+            B: ::ink::env::event::TopicsBuilderBackend<E>,
+        {
+            match self {
+                #topics_builder
+            }
+        }
+     }))
+}
+
+/// The signature topic of an event variant.
+///
+/// Calculated with `blake2b("Event(field1_type,field2_type)")`.
+fn signature_topic(fields: &syn::Fields, event_ident: &syn::Ident) -> syn::LitStr {
```

Comment:
I had an idea of how you can generate a topic for the events with the same field names but different type names (i.e. type aliases).

You can use `TypeId::of::<T>()`. It will infer the type id at the compile type and any aliases and associated types will have the same id. Here is the little demo:
```rust
type Al = u8;

trait Ph {
    type T;
}

struct S;

impl Ph for S {
    type T = u8;
}


fn main() {

    let t1 = TypeId::of::<u8>();
    let t2 = TypeId::of::<Al>();
    let t3 = TypeId::of::<<S as Ph>::T>();
    // all checks out
    assert_eq!(t1, t2);
    assert_eq!(t1, t3);
    assert_eq!(t2, t3);
}
```

TypeId is simply a `u64` value so you can use it to hash. The downside of this approach is that two arbitrary structures that have the same signature (they are identical) but are defined in separate crates will produce different ids.

The only solution to this problem I see is to implement your own version of `TypeId`. Let's call it `EventId`:
* Primitive and built-in types will just use the original id from `TypeId`
* Custom-defined structs which are not events will have to derive `EventId` which will deterministically calculate id based on the fields.
* Any custom-defined structs within that struct should define `EventId` as well.

This is for sure a large additional feature, but this is the only way we can achieve "absolute" determinism in the calculation of the signature topic.

Reply:
Main goals for the signature topic are that it can be determined by the macro at codegen time, and that it is reasonably unique and consistent when used across different contracts.

`TypeId`s are only guaranteed to be the same within the context of a single build of a single binary. An event shared across contracts, or even the same contract but with a different version of Rust or some other variable can and does change the value of this. Using the type ids directly is a non starter, because signature topics must be consistent.

That said, I did consider something similar by piggybacking onto `scale_info::TypeInfo` (which uses `TypeId` behind the scenes). The issue with this and similar approaches is that this information is not available in the macro at codegen time, it is only available at runtime. This is why for `ink!` and `cargo-contract` the metadata step creates a binary to execute and build the metadata.

Another idea I had to be agnostic of the field name would be to have some associated consts for types e.g. `trait TopicName { const NAME: &static str; }` then do some const time concatenation. However this was a little tricky and also the best const hash algorithm I could find was https://docs.rs/xxhash-rust/latest/xxhash_rust/const_xxh3/index.html, and topics are 32 bytes.

So for the sake of consistency with topic hashing (using `blake2` hash algorithm), and just keeping things simple I decided to go with using the type name in the code. 

It is true that two otherwise compatible events, one using a type alias the other not would not have the same signature topic, but I believe that is an acceptable trade off.

Reply:
How about allowing to manually specify the signature topic? If for whatever reason, the exact definition of types can not be used, we can at least allow to specify the topic manually. We are doing that with custom selectors, why not do it with the event signatures?

Reply:
How about allowing to manually specify the signature topic? If for whatever reason, the exact definition of types can not be used, we can at least allow to specify the topic manually. We are doing that with custom selectors, why not do it with the event signatures?

Reply:
It's a good idea, although needs a little thinking about so we can do it as a follow up: https://github.com/paritytech/ink/issues/1838